### DISCUSSION TITLE
Can i using OpenId module authenticate client consuming a GraphQL api

### DISCUSSION THREAD

**Discussion by vitalybrandes** [OrchardCMS/OrchardCore#7263] [Q&A] [answered]
Can i somehow using OpenId module authenticate/authorize client consuming a GraphQL api from external site or i need to make my own module?

**Comment by Skrypt** [comment, marked as answer]
That's what the OpenId module is for. Maybe we need more documentation?

**Reply by vitalybrandes** [reply]
I am not sure that this is documentation issue,
Can I setup clients from code?
As per UI, I can setup only 1 client to authenticate.

**Reply by Skrypt** [reply]
I think we have Scopes? Setup from code I don't know. 
/cc @kevinchalet

**Reply by vitalybrandes** [reply]
Well, scope is still per same 1 client?, No?
What if need to auth with more than 1 remote host,as I collect data from different places.

**Reply by Skrypt** [reply]
If all your hosts are authorized through OpenID then I think it is your javascript client that needs to ask for a token for each of them. I'm not an expert with the OpenID module. Maybe someone else could confirm but I think you can also set the domains that are allowed in the OpenID configurations? I don't want to misslead you on that one so better wait on @kevinchalet

**Reply by vitalybrandes** [reply]
Thank you,
On the same wave, is there any documentation how to use orchardCore graphQl client?

I have external orchard core site with content, and I would like to get some of this data to second site (phisicaly located in different state)  via graphQl query + parameter.
I found that orchard.core.graphql.client exist, but not sure how to build query there for my needs 😟.

**Reply by Skrypt** [reply]
You can initiate an HttpClient request from a Controller Action to keep this more secured or use plain javascript. You will need to allow cross-domain requests by setting up correctly your CORS policies. Basically the GraphQL endpoint needs JSON and returns JSON ...

**Reply by vitalybrandes** [reply]
Thank you.
The CORS are per solution or per module?

**Reply by Skrypt** [reply]
Can be per tenant. There's a PR about this you can read details in the PR. This can be set using simply appsettings.json globally or per tenant. https://github.com/OrchardCMS/OrchardCore/pull/2682

**Reply by vitalybrandes** [reply]
Thanks

**Reply by vitalybrandes** [reply]
Well, i create my own service to do authentication with OpenId server,
but i still curios if there is the service in existing module.

**Reply by kevinchalet** [reply]
The OpenID module has a built-in server you can use to issue tokens than can be validated by the validation feature. To use token authentication in your APIs, use `[Authorize(AuthenticationSchemes = "Api")]`.

Regarding scopes, not sure I got the question right: they are fully supported and they are not tied to a specific client. In the future tho, we'll have a way to configure each app to be able to request each scope or not: https://github.com/OrchardCMS/OrchardCore/issues/1981.

**Reply by vitalybrandes** [reply]
@kevinchalet  Hi,
my original question was regarding OpenId **client** module,
 can i create multiple code based authentications via OpenId **servers** (_to get data from them by API_.)

**Reply by MichaelPetrinolis** [reply]
@vitalybrandes OpenID client does not support multiple servers. It is used to authenticate users from one OpenID provider. You could save external user's token to call an external API that is protected with that IdP. If you need your application to consume external APIs that are protected by different Idps, you should implement your own module to keep the auth tokens for each API and configure the authentication for each one of the HTTP Client that you will use

**Reply by vitalybrandes** [reply]
Thank you!

**Reply by yassinsayed** [reply]
@MichaelPetrinolis  Is there documentation or guide for this approach
